### ISSUE TITLE
IPv4-only indexers disables IPv6 for all indexers

### ISSUE THREAD

**Issue by RomLecat** [Prowlarr/Prowlarr#2157]
### Is there an existing issue for this?

- [X] I have searched the existing open and closed issues

### Current Behavior

Follow-up of #2156.

When an indexer is IPv4-only (in my case, not even bogus IPv6, just no IPv6 at all), it triggers this: https://github.com/Prowlarr/Prowlarr/blob/9f075c09a2c12cedb7e58c9478a7faf4eac8c664/src/NzbDrone.Common/Http/Dispatchers/ManagedHttpDispatcher.cs#L327

Which disables IPv6 on the container.

As explained on #2156, this is an issue because my FlareSolverr runs on IPv6 and Prowlarr needs to have the same IP address.

### Expected Behavior

IPv6 should only be disabled for the indexer that does not support IPv6, not for all indexers.

Also, I think re-evaluate IPv6 connectivity periodically would be better, a single failure in IPv6 results in IPv6 permanently disabled for the lifetime of the application, which seems extreme.

### Steps To Reproduce

1. Add an indexer that only supports IPv4 in a dual-stack (IPv4+IPv6 working environment)
2. Test connectivity to that indexer
3. IPv6 is disabled for all indexers.

### Environment

```markdown
- OS: Debian 12
- Prowlarr: latest nightly
- Docker Install: Kubernetes 1.28.10 in dual-stack mode
- Using Reverse Proxy: Yes
- Browser: Brave
```

### What branch are you running?

Nightly

### Trace Logs?

2024-06-18 12:22:30.8|Debug|Cardigann|Downloading Feed https://www.ygg.re/engine/search?category=all&do=search&order=desc&sort=publish_date&name=&page=50
2024-06-18 12:22:30.9|Trace|ConfigService|Using default config value for 'logindexerresponse' defaultValue:'False'
2024-06-18 12:22:30.9|Trace|RateLimitService|Rate Limit triggered, delaying 'www.ygg.re' for 0.947 sec
2024-06-18 12:22:31.8|Trace|IndexerHttpClient|Req: [GET] https://www.ygg.re/engine/search?category=all&do=search&order=desc&sort=publish_date&name=&page=50
2024-06-18 12:22:31.8|Trace|ConfigService|Using default config value for 'proxyenabled' defaultValue:'False'
2024-06-18 12:22:32.5|Trace|IndexerHttpClient|Res: HTTP/2.0 [GET] https://www.ygg.re/engine/search?category=all&do=search&order=desc&sort=publish_date&name=&page=50: 200.OK (142022 bytes) (631 ms)
2024-06-18 12:22:32.5|Debug|FlareSolverr|CF Protection not detected, returning original response
2024-06-18 12:22:32.5|Trace|Cardigann|Cardigann (yggtorrent): Parsing response
2024-06-18 12:22:32.5|Trace|ConfigService|Using default config value for 'logindexerresponse' defaultValue:'False'
2024-06-18 12:22:32.8|Trace|Cardigann|Cardigann (yggtorrent): Got 50 releases
[...]
2024-06-18 12:22:37.9|Debug|Cardigann|Downloading Feed https://abn.lol/Torrent?SelectedCats=1&SelectedCats=3&SelectedCats=4&SelectedCats=9&YearOperator=%E2%89%A5&RatingOperator=%E2%89%A5&SortOn=Created&SortOrder=desc
2024-06-18 12:22:37.9|Debug|Prowlarr.Http.Authentication.ApiKeyAuthenticationHandler|AuthenticationScheme: API was successfully authenticated.
2024-06-18 12:22:37.9|Trace|ConfigService|Using default config value for 'logindexerresponse' defaultValue:'False'
2024-06-18 12:22:37.9|Trace|Http|Res: 109 [GET] /api/v1/health: 200.OK (8 ms)
2024-06-18 12:22:37.9|Debug|Api|[GET] /api/v1/health: 200.OK (8 ms)
2024-06-18 12:22:37.9|Trace|IndexerHttpClient|Req: [GET] https://abn.lol/Torrent?SelectedCats=1&SelectedCats=3&SelectedCats=4&SelectedCats=9&YearOperator=%E2%89%A5&RatingOperator=%E2%89%A5&SortOn=Created&SortOrder=desc
2024-06-18 12:22:37.9|Trace|ConfigService|Using default config value for 'proxyenabled' defaultValue:'False'
2024-06-18 12:22:38.0|Info|ManagedHttpDispatcher|IPv4 is available: True, IPv6 will be disabled

### Trace Logs have been provided as applicable. Reports may be closed if the required logs are not provided.

- [X] I have read and followed the steps in the wiki link above and provided the required trace logs - the logs contain `trace` - that are relevant and show this issue.

**Comment by Ttfgggf** [comment]
Facing the same issue here as well.

**Comment by RomLecat** [comment]
"Happy eyeballs" have been there for quite a while now, including in major browsers (which is the feature of fallbacking to IPv4 very quicky in case of buggy/not working IPv6. It was quite common a few years ago when adoption started broadly). It makes less sense now that IPv6 issues are *mostly* resolved in most environments.

The specific implementation of the *arr is quite bad, mostly because of the fact that fallback happens only once and is never retried, and that it affects all trackers, which is nonsense.

**Comment by bakerboy448** [comment]
Given that most/many users have ipv6 enabled locally and not for WAN and thus ipv6 connectivity will stall and eventually fail - trying ipv6 every time is not a reasonable approach

**Comment by RomLecat** [comment]
What do you mean by "Ipv6 enabled locally and not for WAN" ?
If you're refering to the link-locale (fe80:: IPv6), it does not have a default route and therefore is using IPv4 by default.

I don't see where "most/many" users having a GUA address does not have proper IPv6 WAN. Either the ISP provides a GUA prefix through SLAAC/DHCPv6 and then it is working, or the ISP does not provide it at all.
"Most/many" users run the *arr suite in Docker, which by default is IPv4-only.

**Comment by ldexterldesign** [comment]
👋

Hope you're well

Thanks for software 😗

I'm currently having to use mobile data instead of a full-fledged ISP, while I finish moving house, and I'm seeing error messages when adding indexers while I'm not using my VPN - my VPN supports ipv6 but my mobile data (ISP) doesn't (ie ipv4-only) so this explains things

It's no big deal to run my VPN 24/7 but without paying for the privilege of a VPN is this resolvable on the prowlarr side?: if yes then I suggest its looked into

[progressive enhance] &/ [graceful degrade] 'n' all that 🤷

Hope to hear back

🤞🙏

PS if i'm straying into a new issue here then feel free to say and I'll delete and recreate

[graceful degrade]: https://en.wikipedia.org/wiki/Fault_tolerance
[progressive enhance]: https://en.wikipedia.org/wiki/Progressive_enhancement

**Comment by kquinsland01** [comment]
+1 for this. Have Prowlarr working in a k8s working well in a dual-stack cluster but when helping friend w/ ipv6 _only_ cluster, the various *arr apps can't communicate. Web interface is up and running on `http://[::]:${port}` and `curl http://readarr.cluster.svc.local` works but as soon as Prowlarr tries to configure an indexer in readarr, vague failure messages about http connection failing :(

**Comment by ldexterldesign** [comment]
> +1 for this. Have Prowlarr working in a k8s working well in a dual-stack cluster but when helping friend w/ ipv6 _only_ cluster, the various *arr apps can't communicate. Web interface is up and running on `http://[::]:${port}` and `curl http://readarr.cluster.svc.local` works but as soon as Prowlarr tries to configure an indexer in readarr, vague failure messages about http connection failing :(

FYI last I heard readarr project was dead so might be why you’re getting errors..? 🤷

**Comment by OdinVex** [comment]
> +1 for this. Have Prowlarr working in a k8s working well in a dual-stack cluster but when helping friend w/ ipv6 _only_ cluster, the various *arr apps can't communicate. Web interface is up and running on `http://[::]:${port}` and `curl http://readarr.cluster.svc.local` works but as soon as Prowlarr tries to configure an indexer in readarr, vague failure messages about http connection failing :(

You'll have to patch and compile for yourself like I do (removed all fallback entirely).

**Comment by Logvin** [comment]
Hey, wanted to drop some context here that might help move this forward.

Sonarr already merged this exact fix. PR [Sonarr#8153](https://github.com/Sonarr/Sonarr/pull/8153) landed in January 2026 and implements a proper Happy Eyeballs algorithm per RFC 8305. The Prowlarr equivalent, PR [#2533](https://github.com/Prowlarr/Prowlarr/pull/2533), has been sitting with zero reviews since October 2025.

The core problem here is that a single IPv4-only indexer (one that just lacks AAAA records... that's normal, not broken) permanently disables IPv6 for the entire process. That's a static boolean acting as a global kill switch. It conflates "this host doesn't support IPv6" with "my system's IPv6 is broken." Those are very different things.

Happy Eyeballs isn't some niche solution. It's RFC 8305, and the vast majority of network-aware software already implements it:

- **curl** uses it with a 200ms connection attempt delay
- **Chrome, Firefox, Safari** all implement it per-connection
- **Jellyfin** added it for their .NET HttpClient ([jellyfin#5735](https://github.com/jellyfin/jellyfin/pull/5735))
- **.NET 11** is finally adding native support via `ConnectAlgorithm.Parallel`

The last company I worked for was a very large ISP with a very tiny IPv4 space; we heavily utilized pure IPv6, and did not issue IPv4 addresses to our customers, so I got pretty good at troubleshooting IPv4/IPv6 translation issues. IPv6 reachability varies per destination, not per source. A global flag is always the wrong approach because one unreachable host says nothing about whether your next connection will work over v6. Happy Eyeballs handles this correctly by racing connections per-host with a 250ms delay. It's fast, it's fair, and it doesn't break anything.

The fix is written, tested (222 lines of tests), and proven in Sonarr. I'm building a test lab to help validate the behavior in PR #2533 and will update here when I have results. I'd love to see it get some review attention in the meantime.

Could we also update the labels on this one?

- Remove `Priority: Low` and add `Priority: Medium`. A proven fix exists and is merged in a sibling project. This isn't speculative.
- Remove `Status: Help Wanted`. Help arrived. PR #2533 is ready for review.
- Remove `Type: Enhancement`. This is a bug, not a feature request. The current behavior is incorrect.
- Keep `Type: Bug`.

**Comment by OdinVex** [comment]
> Hey, wanted to drop some context here that might help move this forward.
> 
> Sonarr already merged this exact fix. PR [Sonarr#8153](https://github.com/Sonarr/Sonarr/pull/8153) landed in January 2026 and implements a proper Happy Eyeballs algorithm per RFC 8305. The Prowlarr equivalent, PR [#2533](https://github.com/Prowlarr/Prowlarr/pull/2533), has been sitting with zero reviews since October 2025.
> 
> The core problem here is that a single IPv4-only indexer (one that just lacks AAAA records... that's normal, not broken) permanently disables IPv6 for the entire process. That's a static boolean acting as a global kill switch. It conflates "this host doesn't support IPv6" with "my system's IPv6 is broken." Those are very different things.
> 
> Happy Eyeballs isn't some niche solution. It's RFC 8305, and the vast majority of network-aware software already implements it:
> 
>     * **curl** uses it with a 200ms connection attempt delay
> 
>     * **Chrome, Firefox, Safari** all implement it per-connection
> 
>     * **Jellyfin** added it for their .NET HttpClient ([jellyfin#5735](https://github.com/jellyfin/jellyfin/pull/5735))
> 
>     * **.NET 11** is finally adding native support via `ConnectAlgorithm.Parallel`
> 
> 
> The last company I worked for was a very large ISP with a very tiny IPv4 space; we heavily utilized pure IPv6, and did not issue IPv4 addresses to our customers, so I got pretty good at troubleshooting IPv4/IPv6 translation issues. IPv6 reachability varies per destination, not per source. A global flag is always the wrong approach because one unreachable host says nothing about whether your next connection will work over v6. Happy Eyeballs handles this correctly by racing connections per-host with a 250ms delay. It's fast, it's fair, and it doesn't break anything.
> 
> The fix is written, tested (222 lines of tests), and proven in Sonarr. I'm building a test lab to help validate the behavior in PR [#2533](https://github.com/Prowlarr/Prowlarr/pull/2533) and will update here when I have results. I'd love to see it get some review attention in the meantime.
> 
> Could we also update the labels on this one?
> 
>     * Remove `Priority: Low` and add `Priority: Medium`. A proven fix exists and is merged in a sibling project. This isn't speculative.
> 
>     * Remove `Status: Help Wanted`. Help arrived. PR [Use Happy Eyeballs for HTTP socket address selection #2533](https://github.com/Prowlarr/Prowlarr/pull/2533) is ready for review.
> 
>     * Remove `Type: Enhancement`. This is a bug, not a feature request. The current behavior is incorrect.
> 
>     * Keep `Type: Bug`.

I reiterate, Happy Eyeballs should not exist. Sonarr's change was not a fix, at all. IPv6, then IPv4. If users want to prefer a stack, fine. Just because IPv6 resolution may fail once doesn't mean to try only trying IPv4 afterward. This has nothing to do with the fact one disables for all, it's the entire disablement to begin with. Your posts read like AI. >_> Code in general is just getting worse and will keep getting worse.

**Comment by Logvin** [comment]
>I reiterate, Happy Eyeballs should not exist

That is your opinion. When I look at situations like this I start with "what is the right way to handle this?"... if there are existing best practices available, I always try and use that first. HE is the best practice.

HE handles your issue perfectly... when Prowlarr makes a connection attempt it will try IPv6, if that fails within 250ms it goes for IPv4. Nothing is disabled for one or all indexers at all. Lets say you have an indexer running a dual stack IPv4/IPv6 and their IPv6 goes down for an hour... implementing HE would allow Prowlarr to just keep on trucking along, and once IPv6 comes back it will shift back to that. There are no variables needing to be stored in Prowlarr about which IP versions indexers use or work - each attempt  verifies independently for each session. This is the best practice design.

>Your posts read like AI. >_> Code in general is just getting worse and will keep getting worse.

Yes, I absolutely use AI when I work on issues. I disclose that with all of my PRs and in my Github profile. I am a bit neurodiverse, and when I write I am often told my tone is aggressive. I have found that AI helps me get the words out of my head and organized in a way that works for others better than when I do it alone. They are my words.

I saw a great comment a few weeks ago and saved it, I think it applies here.

# AD HOMINEM (ATTACK THE MESSENGER)

---

## What It Is

Attacking the person making the argument instead of the argument itself. If you can't debunk the message, go after the messenger.

---

## Everyday Example

> "You failed math in high school, so your budget spreadsheet must be wrong."

Whether you failed math has nothing to do with whether the spreadsheet adds up. Check the numbers.

---

## How It Applies Here

> "This looks like AI wrote it."

Even if every word were AI-generated, that wouldn't change whether the information in the post is accurate. Criticizing *how* something was written is not the same as showing it is *wrong*. Engage with the content, not the author.

This comment ignores the actual substance of the post and targets the method of communication instead. That's the definition of an ad hominem: it's a rhetorical move, not a rebuttal.

---

PS: My AI bot said I should not lecture you about this because it comes off as aggressively toned. I told the clanker to can it.

>Code in general is just getting worse and will keep getting worse.

I disagree. I see such amazing things happening in the software development space! I'm not a software dev by trade... I am an infrastructure, security, network, and troubleshooter. I have a process that I use to troubleshoot issues that I developed over the past few decades of work. I have been diving into contributing to open source software because I noticed that my skills are really needed here... people write a fix, but often those fixes are abandoned because validating them is a burden. Well... I'm really good at developing test plans and validating solutions.

If your goal here is to get Prowlarr to stop disabling IPv6 for all indexers, this is the fix. I get that it is not your preferred way to fix it, but it is the best practice and widely used.

**Comment by OdinVex** [comment]
> That is your opinion. When I look at situations like this I start with "what is the right way to handle this?"... if there are existing best practices available, I always try and use that first. HE is the best practice.
> 
> HE handles your issue perfectly...

IPv4 was stretched too thin for years and should already have been phased out. There's zero reason for HE to exist. Maintaining split DNS that strips IPv4 records and synthesizes IPv6 to fix bugs like this is a hassle, to protect legacy addressing.

> Yes, I absolutely use AI when I work on issues. I disclose that with all of my PRs and in my Github profile. I am a bit neurodiverse, and when I write I am often told my tone is aggressive. I have found that AI helps me get the words out of my head and organized in a way that works for others better than when I do it alone. They are my words.

They're not, but whatever.

> I saw a great comment a few weeks ago and saved it, I think it applies here.
> # AD HOMINEM (ATTACK THE MESSENGER)
> ...Removed for brevity of post...

"Oh no, he doesn't like my AI-generated post and can tell its AI-generated. Attack him for pointing out how it sounds with a straw man argument. <insert some AI suck-off>" It's about ethics and functionality. No, AI isn't ethical for the same reasons we don't use the research Nazis did in their day, unethical means of production. Disclosing AI usage is only part of the ethics, making sure they're not unethical ones (you mention Claude in your profile) is the other half. As for functionality, AI slop vulnerabilities.

> Even if every word were AI-generated, that wouldn't change whether the information in the post is accurate. Criticizing _how_ something was written is not the same as showing it is _wrong_. Engage with the content, not the author.

The argument for HE is wrong to begin with. I'm wondering if you even know why it is called Happy Eyeballs. It's because most end users don't like to wait for content to visibly show so both IPv6 and IPv4 are tried in hopes of showing something *faster*, eg a brute-force solution rather than intelligent one: AAAA exists, use it.

It isn't an ad hominem situation because I'm against both the position and the poster's conveyance of position. But I guess AI didn't tell you that.

> PS: My AI bot said I should not lecture you about this because it comes off as aggressively toned. I told the clanker to can it.

It was aggressive to start, that was the easy pickup of AI (the condescension without anything to even stand on).

> I disagree. I see such amazing things happening in the software development space! I'm not a software dev by trade...

That's probably the problem, then. I'll try to throw it a bone: Ignoring the ethics and the random-off issues AI can have, sure, it can be *helpful* but never trusted, nor relied upon, but that's not what my post was about (as explained in this one since AI didn't seem to tell you what it was about). To reiterate it to make sure it's understood, my points were HE not being needed and your using AI to help in some way didn't give it standing for defense.

> If your goal here is to get Prowlarr to stop disabling IPv6 for all indexers, this is the fix. I get that it is not your preferred way to fix it, but it is the best practice and widely used.

My goal is to stop Prowlarr from trying to be "helpful" because it's unhelpful (and rather permanently so, quite disastrously in dual-stack setups). Like AI, it's a Clippy move. Disabling IPv6 ever (in a system with it enabled, as they should be these days) is an issue. The entire point of specifying an AAAA record means the domain administrator(s) intend for traffic reception at an address. If they're legacy-only, A. If an ISP is still legacy and only supports IPv4 then those clients get IPv4 resolution, so there's no reason to implement any kind of HE. At this point, we're just going to have to disagree. Too experienced to waste further time on this.

**Comment by Logvin** [comment]
Got some results from the test lab.

I built a Docker environment with mock Torznab indexers on specific address families and controlled `/etc/hosts` entries. Full test plan and methodology [here](https://github.com/Logvin/lab-notes/tree/main/test-labs/prowlarr-ipv6-happy-eyeballs).

**Baseline (v2.3.4.5307, current develop):**

This is worse than I expected. IPv6 was globally disabled *on startup*, before any indexer was even contacted. The update check triggered the kill switch:

```
[Info] ManagedHttpDispatcher: IPv4 is available: True, IPv6 will be disabled
```

| Indexer | DNS | Connections | Address Family | Result |
|---------|-----|-------------|----------------|--------|
| v4only | A record only | 2 | IPv4 | OK |
| v6only | AAAA record only | 0 | N/A | **Unreachable** |
| dualstack | A + AAAA | 2 | IPv4 only | Degraded |

The v6-only indexer threw `HttpRequestException: No data available` from `ManagedHttpDispatcher.attemptConnection`. It tried to connect via IPv4 to a host that only has an AAAA record. Nothing it can do.

**PR #2533 (Happy Eyeballs fix, built from commit 0299647):**

No "IPv6 will be disabled" message. No global flag. Every connection races both address families independently.

| Indexer | DNS | Connections | Address Family | Result |
|---------|-----|-------------|----------------|--------|
| v4only | A record only | 2 | IPv4 | OK |
| v6only | AAAA record only | 2 | **IPv6** | **OK** |
| dualstack | A + AAAA | 2 | **IPv6 preferred** | OK |

The dualstack result is worth calling out... baseline forced it to IPv4 even though a perfectly good AAAA record existed. With Happy Eyeballs, IPv6 won the race. That's the expected behavior per RFC 8305.

Tested on two independent systems (macOS Docker Desktop + Ubuntu 24.04 bare metal) with identical results. A [reproducible GitHub Actions workflow](https://github.com/Logvin/lab-notes) is available if anyone wants to validate independently.

PR #2533 fixes the issue cleanly. I'd love to see it get merged.

**AI Disclosure:** This test lab was developed with Claude Code as a pair programming tool. I directed the approach, designed the test methodology, reviewed all results, and validated the findings independently. I'm the author, Claude is my editor.

**Comment by OdinVex** [comment]
A flag to disable everything relating to it would allow anyone to control their own installation, could make everyone happy, too (with your PR).
